miniapps/adjoint/cvsRoberts_ASAi_dns.cpp

```diff
@@ -0,0 +1,385 @@
+
+      done = (t >= t_final - 1e-8*dt);
+
+      if (done)
+      {
+	cvodes->PrintInfo();
+      }
+   }
+
+   cout << "Final Solution: " << t << endl;
+   u.Print();
+
+   q = 0.;
+   cout << " Final Quadrature " << endl;
+   cvodes->EvalQuadIntegration(t, q);
+   q.Print();
+
+   // 7. Solve the adjoint problem at different points in time
+
+   // Create the adjoint solution vector
+   Vector w(3);
+   w=0.;
+   double TBout1 = 40.;
+   Vector dG_dp(3);
+   dG_dp=0.;
+   if (cvodes)
+   {
+      t = t_final;
+      adv.SetTime(t);
+      cvodes->InitB(adv);
+      cvodes->InitQuadIntegrationB(dG_dp, 1.e-6, 1.e-6);
+
+      // Results at time TBout1
+      double dt_real = max(dt, t - TBout1);
+      cvodes->StepB(w, t, dt_real);
+      cout << "t: " << t << endl;
+      cout << "w:" << endl;
+      w.Print();
+
+      cvodes->GetForwardSolution(t, u);
+      cout << "u:" << endl;
+      u.Print();
+
+      // Results at T0
+      dt_real = max(dt, t - 0.);
+      cvodes->StepB(w, t, dt_real);
+      cout << "t: " << t << endl;
+      cout << "w:" << endl;
+
+      cvodes->GetForwardSolution(t, u);
+      w.Print();
+      cout << "u:" << endl;
+      u.Print();
+
+      // Evaluate Sensitivity
+      cvodes->EvalQuadIntegrationB(t, dG_dp);
+      cout << "dG/dp:" << endl;
+      dG_dp.Print();
+
+   }
+
+   // 8. Free the used memory.
+   delete cvodes;
+
+   return 0;
+}
+
+// cvsRoberts_ASAi_dns rate equation
+void RobertsTDAOperator::Mult(const Vector &x, Vector &y) const
+{
+   y[0] = -p_[0]*x[0] + p_[1]*x[1]*x[2];
+   y[2] = p_[2]*x[1]*x[1];
+   y[1] = -y[0] - y[2];
+}
+
+// cvsRoberts_ASAi_dns quadrature rate equation
+void RobertsTDAOperator::QuadratureIntegration(const Vector &y,
+					       Vector &qdot) const
+{
+   qdot[0] = y[2];
+}
+
+// cvsRoberts_ASAi_dns adjoint rate equation
+void RobertsTDAOperator::AdjointRateMult(const Vector &y, Vector & yB,
+                                      Vector &yBdot) const
+{
+   double l21 = (yB[1]-yB[0]);
+   double l32 = (yB[2]-yB[1]);
+   double p1 = p_[0];
+   double p2 = p_[1];
+   double p3 = p_[2];
+   yBdot[0] = -p1 * l21;
+   yBdot[1] = p2 * y[2] * l21 - 2. * p3 * y[1] * l32;
+   yBdot[2] = p2 * y[1] * l21 - 1.0;
+}
+
+// cvsRoberts_ASAi_dns quadrature sensitivity rate equation
+void RobertsTDAOperator::QuadratureSensitivityMult(const Vector &y,
+						   const Vector &yB,
+						   Vector &qBdot) const
+{
+   double l21 = (yB[1]-yB[0]);
+   double l32 = (yB[2]-yB[1]);
+   double y23 = y[1] * y[2];
+
+   qBdot[0] = y[0] * l21;
+   qBdot[1] = -y23 * l21;
+   qBdot[2] = y[1]*y[1]*l32;
+}
+
+// cvsRoberts_ASAi_dns implicit solve setup for adjoint
+int RobertsTDAOperator::SUNImplicitSetupB(const double t, const Vector &y,
+					  const Vector &yB,
+					  const Vector &fyB, int jokB,
+					  int *jcurB, double gammaB)
+{
+
+   // M = I- gamma J
+   // J = dfB/dyB
+   // Let's create a SparseMatrix and fill in the entries since this example
+   // doesn't contain finite elements
```

Comment:
This comment doesn't sound right. I see your point but would suggest reformulating it.

Reply:
I can just take out the comment if you would like. Or something like "no integrator is being used".

Reply:
It is an ODE test problem in R^n so I think its fine to remove the comment.